doc/hacking.rst

```diff
@@ -316,3 +316,179 @@ For automatically updated builds, you can install sphinx-autobuild, and run
+
+- Every ``.ml`` file must have a corresponding ``.mli``. The only exception to
+  this rule is ``.ml`` files with only type definitions.
+
+- Do not write ``.mli`` only modules. They offer no advantages to ``.ml``
+  modules with type definitions and one cannot define exceptions in ``.mli``
+  only modules
+
+- Every module should have toplevel documentation that describes the module
+  briefly. This is a good place to discuss its purpose, invariants, etc.
+
+- Keep interfaces short & sweet. The less functions, types, etc. there are, the
+  easier it is for users to understand, use, and ultimately modify the
+  interface correctly. Instead of creating elaborate interfaces with the hope
+  of future-proofing every use case, embrace change and make it easier to throw
+  out or replace the interface.
+
+  Ideally the interface should have one obvious way to use it. A particularly
+  annoying violator of this principle is the "logic-less chain of functions"
+  helper. For example:
+
+.. code:: ocaml
+
+   let foo t = bar t |> baz
+
+  If ``bar`` and ``baz`` are already public, then there's no need to add yet
+  another helper to save the caller a line of code.
+
+- Define bindings as close to their use site as possible. When they're far
+  apart, reading code requires scrolling and IDE tools to understand the code.
+
+Bad:
+
+.. code:: ocaml
+
+   let dir = .. in
+   (* 50 odd lines or so that don't use [dir] *)
+   f dir
+
+Good:
+
+.. code:: ocaml
+
+  let dir = .. in
+  f dir
+
+- A corollary to the previous guideline: keep the scope of bindings as small as
+  possible.
+
+Bad:
+
+.. code:: ocaml
+
+   let x1 = f foo in let x2 = f bar in
+   let y1 = g foo in let y2 = g bar in
+   let dx = x2 -. x1 in
+   let dy = y2 -. y1 in
+   dx^2 +. dy^2
+
+Good:
+
+.. code:: ocaml
+
+   let dx =
+     let x1 = f foo in let x2 = f bar in
+     x2 -. x1
+   in
+   let dy =
+     let y1 = g foo in let y2 = g bar in
+     y2 -. y1
+   in
+   dx^2 +. dy^2
+
+- Prefer ``Code_error.raise`` instead of ``assert false``. The reader often has
+  no idea what invariant is broken by the ``assert false``. Kindly describe it
+  to the reader in the error message.
+
+- Avoid meangingless names like `x`, `a`, `b`, `f`. Try to find a more
+  descriptive name or just inline it altogether.
+
+- If a module ``Foo`` has a module type ``Foo.S`` and you'd like to avoid
+  repeating its definition in the implementation and the signature, introduce
+  an ``.ml``-only module ``Foo_intf`` and write the ``S`` only once in there.
+
+- Instead of introducing a type ``foo``, consider introducing a module ``Foo``
+  with a type ``t``. This is often the place to put functions related to
+  ``foo``.
+
+- Avoid optional arguments. They increase brevity at the expense of readability
+  and are annoying to grep. Further more, they encourage callers not to think
+  at all about these optional arguments even if they often should.
+
+- Stage functions explicitly with the ``Stage`` module.
```

Comment:
Do we actually use this idiom? I don't remember seeing it. I kind of mostly dislike it, particularly because people tend to be pretty random/unprintipled about where it's used.

Reply:
We do use it and not using it has caused some serious performance regressions in the past. I'm pretty sure we also inherited it from janestreet https://blog.janestreet.com/making-staging-explicit/